plan/relations.go

```diff
@@ -564,6 +690,44 @@ func (j *JoinRel) ToProtoPlanRel() *proto.PlanRel {
 	}
 }
 
+func (j *JoinRel) GetInputs() []Rel {
+	return []Rel{j.left, j.right}
+}
+
+func (j *JoinRel) Copy(newInputs ...Rel) (Rel, error) {
+	if len(newInputs) != 2 {
+		return nil, substraitgo.ErrInvalidInputCount
+	}
+	join := *j
+	join.left = newInputs[0]
+	join.right = newInputs[1]
+	return &join, nil
+}
+
+func (j *JoinRel) CopyWithExpressionRewrite(rewriteFunc func(expr.Expression) (expr.Expression, error), newInputs ...Rel) (Rel, error) {
+	if len(newInputs) != 2 {
+		return nil, substraitgo.ErrInvalidInputCount
+	}
+	currExprs := []expr.Expression{j.expr, j.postJoinFilter}
+	exprs := make([]expr.Expression, len(currExprs))
+	for i, e := range currExprs {
+		newExpr, err := rewriteFunc(e)
+		if err != nil {
+			return nil, err
+		}
+		exprs[i] = newExpr
+	}
+	if areExpressionsEqual(exprs, currExprs) && areRelsEqual(newInputs, j.GetInputs()) {
+		return nil, nil
+	}
+	join := *j
+	join.left = newInputs[0]
+	join.right = newInputs[1]
+	join.expr = exprs[0]
+	join.postJoinFilter = exprs[1]
```

Comment:
Could we do

```go
join := j.Copy(newInputs...)
join.expr, join.postJoinFilter = exprs[0], exprs[1]
```

instead?

Reply:
Did this partially. Using j.Copy asks for a type cast to JoinRel.